app/controllers/bookings_controller.rb

```diff
@@ -13,32 +13,58 @@ def create
     @booking = Booking.new(booking_params)
     # assign space_id, user_id and price_per_day to the booking
     @booking.space = @space
-    # TODO => how to ask user to login/sign up before booking
+    # assign booking user to current_user
     @booking.user = current_user
     # TODO => how to make @booking.price_per_day a static?
     @booking.price_per_day = @space.price_per_day
-    # TODO => loop to check that the space is available
+    # current_user date input in form
+    requested_arrival_date =@booking.start_date
+    requested_departure_date = @booking.end_date
+    # all existing bookings for that space
+    existing_bookings = @space.bookings
+    # for each [start_date, end_date] in space bookings
+    # arrival_date should be before start_date -> < start_date
+    # and departure_date should be after end_date -> > end_date
+
+    # @date_start.between?(arival_date, leave_date) || @date_end.between?(arival_date, leave_date)
     # at the dates entered in the form
+
     # if available
         # => post book
         # => direct to bookings dashboard
     # if not available
         # => stay on same page
         # => display message
     #raise
-    if @booking.save
+    if is_available?(existing_bookings, requested_arrival_date, requested_departure_date)
+      @booking.save
       flash[:notice] = "Your booking is confirmed!"
-      redirect_to space_path(@space)
+      redirect_to space_booking_path(@space, @booking)
     else
       flash[:notice] = "Dates not available :("
       redirect_to space_path(@space)
     end
   end
 
+  def show
+    @space = Space.find(params[:space_id])
+    @booking = Booking.find(params[:id])
+  end
+
   private
 
   def booking_params
     params.require(:booking).permit(:start_date, :end_date)
   end
 
+  def is_available?(existing_bookings, requested_arrival_date, requested_departure_date)
+      existing_bookings.each do |booking|
+      if requested_arrival_date.between?(booking.start_date, booking.end_date) || requested_departure_date.between?(booking.start_date, booking.end_date)
+        return false
+      else
+        return true
```

Comment:
@Bouchka a little bit of an issue with the logic here. Based on this condition, it will only check for the first booking in existing_bookings because the return will break out of the method on the first loop. Here's my suggestion:

```suggestion
      existing_bookings.all do |booking|
        if requested_arrival_date.between?(booking.start_date, booking.end_date) || requested_departure_date.between?(booking.start_date, booking.end_date)
        return false
      else
        return true
```
The .all iterator will check that for all bookings the value returned is true and only then the method will return true. if any of the iterations returns a false, then the method will return a false